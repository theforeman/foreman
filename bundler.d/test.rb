group :test do
  gem 'mocha', '~> 1.1', :require => false
  gem 'simplecov', '~> 0.9'
  gem 'spork-minitest', '0.0.3'
  gem 'single_test', '~> 0.6'
  gem 'minitest', '5.1.0'
  gem 'minitest-spec-rails' #, '~> 4.7'
  #gem 'ci_reporter', '2.0.0', :require => false
  gem 'ci_reporter_minitest', :require => false
  gem 'capybara', '~> 2.0'
  gem 'database_cleaner', '~> 1.3'
  gem 'launchy', '~> 2.4'
  gem 'spork', '~> 0.9'
  gem 'factory_girl_rails', '~> 4.5', :require => false
  gem 'rubocop-checkstyle_formatter', '~> 0.1'
  gem "poltergeist"
end
