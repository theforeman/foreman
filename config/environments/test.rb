# Track deprecation warnings in test environment as early as possible, but pause processing of
# deprecations until all plugins are registered (prior to the finisher_hook initializer) to ensure
# the whitelist is fully configured. This is done in the after_initialize block below.
ASDeprecationTracker.pause!

Foreman::Application.configure do
  # Settings specified here will take precedence over those in config/application.rb.

  # The test environment is used exclusively to run your application's
  # test suite. You never need to work with it otherwise. Remember that
  # your test database is "scratch space" for the test suite and is wiped
  # and recreated between test runs. Don't rely on the data there!
  config.cache_classes = true

  # Do not eager load code on boot. This avoids loading your whole application
  # just for the purpose of running a single test. If you are using a tool that
  # preloads Rails for running tests, you may have to set it to true.
  config.eager_load = false

  # Configure static asset server for tests with Cache-Control for performance.
  config.serve_static_files   = true
  config.static_cache_control = 'public, max-age=3600'

  # Adds additional error checking when serving assets at runtime.
  # Checks for improperly declared sprockets dependencies.
  # Raises helpful error messages.
  config.assets.raise_runtime_errors = true

  # Show full error reports and disable caching.
  config.consider_all_requests_local       = true
  config.action_controller.perform_caching = false

  # Raise exceptions instead of rendering exception templates.
  config.action_dispatch.show_exceptions = false

  # Disable request forgery protection in test environment.
  config.action_controller.allow_forgery_protection = false

  # Tell Action Mailer not to deliver emails to the real world.
  # The :test delivery method accumulates sent emails in the
  # ActionMailer::Base.deliveries array.
  config.action_mailer.delivery_method = :test
  config.action_mailer.perform_deliveries = true
  ActionMailer::Base.deliveries.clear

  # Use SQL instead of Active Record's schema dumper when creating the test database.
  # This is necessary if your schema can't be completely dumped by the schema dumper,
  # like if you have constraints or database-specific column types
  # config.active_record.schema_format = :sql

  # Should ANSI color codes be used when logging information
  config.colorize_logging = Foreman::Logging.config[:colorize]

  # Print deprecation notices to the stderr.
  config.active_support.deprecation = :stderr

  # Raise exception on mass assignment of unfiltered parameters
  config.action_controller.action_on_unpermitted_parameters = :strict

  # Use separate cache stores for parallel_tests
  config.cache_store = :file_store, Rails.root.join("tmp", "cache", "paralleltests#{ENV['TEST_ENV_NUMBER']}")

  # Enable automatic creation/migration of the test DB when running tests
  config.active_record.maintain_test_schema = true

  # Randomize the order test cases are executed.
  config.active_support.test_order = :random

  config.webpack.dev_server.enabled = false

  # Whitelist all plugin engines by default from raising errors on deprecation warnings for
  # compatibility, allow them to override it by adding an ASDT configuration file.
  config.after_initialize do
    Foreman::Plugin.all.each do |plugin|
      unless File.exist?(File.join(plugin.path, 'config', 'as_deprecation_whitelist.yaml'))
        ASDeprecationTracker.whitelist.add(engine: plugin.id.to_s.gsub('-', '_'))
      end
    end
    ASDeprecationTracker.resume!
  end
end
