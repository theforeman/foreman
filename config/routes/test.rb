if Rails.env.test?
  Foreman::Application.routes.draw do
    resources :testable, :only => :index
    resources :testable_strong_params, :only => :create
    resources :my_hosts, :only => :create
    resources :testable_no_params, :only => :create

    namespace :api do
      namespace :v2 do
        resources :testable, :only => [:create, :index]
      end

      resources :testable, :only => :index do
        get :raise_error, :on => :collection
      end
    end
  end
end
