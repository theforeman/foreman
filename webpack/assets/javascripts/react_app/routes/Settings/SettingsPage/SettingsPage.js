import React from 'react';
import PropTypes from 'prop-types';

import { translate as __ } from '../../../common/I18n';
import PageLayout from '../../common/PageLayout/PageLayout';
import { SETTINGS_SEARCH_PROPS } from '../constants';
import SettingsPageContent from './components/SettingsPageContent';

const SettingsPage = ({
  pageParams,
  isLoading,
  hasData,
  fetchAndPush,
  hasError,
  errorMsg,
  groupedSettings,
}) => (
  <PageLayout
    header={__('Settings')}
    searchable
    searchProps={SETTINGS_SEARCH_PROPS}
    searchQuery={pageParams.search}
    isLoading={isLoading && hasData}
    onSearch={search => fetchAndPush({ search })}
    onBookmarkClick={search => fetchAndPush({ search })}
  >
    <SettingsPageContent
      groupedSettings={groupedSettings}
      hasData={hasData}
      hasError={hasError}
      isLoading={isLoading}
      errorMsg={errorMsg}
    />
  </PageLayout>
);

SettingsPage.propTypes = {
  pageParams: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired,
  hasData: PropTypes.bool.isRequired,
  fetchAndPush: PropTypes.func.isRequired,
  hasError: PropTypes.bool.isRequired,
  errorMsg: PropTypes.object.isRequired,
  groupedSettings: PropTypes.object.isRequired,
};

export default SettingsPage;
