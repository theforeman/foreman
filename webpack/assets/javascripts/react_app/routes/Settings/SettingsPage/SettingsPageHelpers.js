export const demodulize = title => title.split('::')[1];

export const stickCategory = sticky => categories =>
  categories.includes(sticky)
    ? [sticky, ...categories.filter(item => item !== sticky)]
    : categories;

export const stickGeneralFirst = stickCategory('Setting::General');

export const groupSettings = settings =>
  settings.reduce((memo, setting) => {
    if (!memo[setting.category]) {
      memo[setting.category] = [setting];
      return memo;
    }
    memo[setting.category].push(setting);
    return memo;
  }, {});
