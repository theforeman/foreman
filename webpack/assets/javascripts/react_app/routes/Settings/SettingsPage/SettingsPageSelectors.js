import Immutable from 'seamless-immutable';

import { groupSettings } from './SettingsPageHelpers';

export const selectSettingsPage = state => state.settingsPage.pageContent;

export const selectSettings = state =>
  selectSettingsPage(state).results || Immutable([]);

export const selectHasError = state => selectSettingsPage(state).hasError;
export const selectIsLoading = state => selectSettingsPage(state).isLoading;
export const selectHasData = state => selectSettingsPage(state).hasData;
export const selectErrorMsg = state => selectSettingsPage(state).message;

export const selectGroupedSettings = state =>
  groupSettings(selectSettings(state));

export const selectSettingGroup = state => group =>
  selectGroupedSettings(state)[group];
