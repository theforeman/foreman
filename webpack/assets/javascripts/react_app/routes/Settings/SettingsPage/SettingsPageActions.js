import URI from 'urijs';

import history from '../../../history';
import { API } from '../../../redux/API';

import { deepPropsToCamelCase } from '../../../common/helpers';

import {
  SETTINGS_PAGE_DATA_RESOLVED,
  SETTINGS_PAGE_DATA_FAILED,
  SETTINGS_PAGE_CLEAR_ERROR,
  SETTINGS_PAGE_SHOW_LOADING,
  SETTINGS_API_PATH,
  SETTINGS_PATH,
} from '../constants';

import { selectHasError } from './SettingsPageSelectors';

export const initializeSettings = initialParams => dispatch => {
  dispatch(fetchSettings(initialParams));

  if (!history.action === 'POP') {
    history.replace({
      pathname: SETTINGS_PATH,
      search: URI()
        .search(initialParams)
        .query(),
    });
  }
};

const transformParams = params => ({ search: params.search });

export const fetchSettings = (params, url = SETTINGS_API_PATH) => async (
  dispatch,
  getState
) => {
  dispatch({ type: SETTINGS_PAGE_SHOW_LOADING });

  if (selectHasError(getState())) {
    dispatch({ type: SETTINGS_PAGE_CLEAR_ERROR });
  }

  try {
    const { data } = await API.get(url, {}, transformParams(params));

    const transformedResponse = deepPropsToCamelCase(data);

    dispatch({
      type: SETTINGS_PAGE_DATA_RESOLVED,
      payload: {
        ...transformedResponse,
        hasData: transformedResponse.subtotal > 0,
        isLoading: false,
      },
    });
  } catch (error) {
    const { response } = error;
    let baseType = response.status;
    let baseText = response.statusText;
    if (response && response.data) {
      const { type, text } = response.data;
      baseType = type || baseType;
      baseText = text || baseText;
    }
    dispatch({
      type: SETTINGS_PAGE_DATA_FAILED,
      payload: {
        message: {
          type: baseType,
          text: baseText,
        },
        isLoading: false,
      },
    });
  }
};

export const fetchAndPush = params => (dispatch, getState) => {
  const query = buildQuery(params);
  dispatch(fetchSettings(query));
  history.push({
    pathname: SETTINGS_PATH,
    search: URI()
      .search(query)
      .query(),
  });
};

const buildQuery = params => (params.search ? params : {});
