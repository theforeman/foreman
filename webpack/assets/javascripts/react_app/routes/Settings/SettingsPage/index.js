import Immutable from 'seamless-immutable';
import { compose, bindActionCreators, combineReducers } from 'redux';
import { connect } from 'react-redux';
import URI from 'urijs';

import SettingsPage from './SettingsPage';
import { deepPropsToCamelCase } from '../../../common/helpers';

import * as actions from './SettingsPageActions';

import { callOnMount, callOnPopState } from '../../../common/HOC';

import withDataReducer from '../../common/reducerHOC/withDataReducer';

import testEmailReducer from './components/TestEmail/TestEmailReducer';

import {
  selectSettings,
  selectGroupedSettings,
  selectIsLoading,
  selectHasError,
  selectHasData,
  selectErrorMsg,
} from './SettingsPageSelectors';

const mapStateToProps = (state, ownProps) => ({
  settings: selectSettings(state),
  pageParams: paramsFromHistory(ownProps.history),
  groupedSettings: selectGroupedSettings(state),
  isLoading: selectIsLoading(state),
  hasData: selectHasData(state),
  hasError: selectHasError(state),
  errorMsg: selectErrorMsg(state),
});

export const initialState = Immutable({
  results: [],
});

const extendReducer = (state, type, payload) => {
  switch (type) {
    case 'SETTINGS_FORM_SUBMITTED': {
      return state.set(
        'results',
        state.results.map(setting =>
          payload.data && payload.data.id === setting.id
            ? deepPropsToCamelCase(payload.data)
            : setting
        )
      );
    }
    default:
      return state;
  }
};

export const reducers = {
  settingsPage: combineReducers({
    pageContent: withDataReducer('SETTINGS_PAGE', initialState, extendReducer),
    testEmail: testEmailReducer,
  }),
};

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

const paramsFromHistory = history => {
  const baseParams = { search: '' };

  if (!history || !history.location) {
    return baseParams;
  }

  const parsedSearch = URI(history.location.search).search(true);

  return parsedSearch.search ? { search: parsedSearch.search } : baseParams;
};

const onMount = props => props.initializeSettings(props.pageParams);

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  callOnMount(onMount),
  callOnPopState(onMount)
)(SettingsPage);
