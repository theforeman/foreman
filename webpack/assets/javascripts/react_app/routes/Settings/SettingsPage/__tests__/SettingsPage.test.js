import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import SettingsPage from '../SettingsPage';

const fixtures = {
  'should render': {
    pageParams: { search: '' },
    isLoading: true,
    hasData: false,
    fetchAndPush: () => {},
    hasError: false,
    errorMsg: {},
    groupedSettings: {},
  },
};

describe('SettingsPage', () =>
  testComponentSnapshotsWithFixtures(SettingsPage, fixtures));
