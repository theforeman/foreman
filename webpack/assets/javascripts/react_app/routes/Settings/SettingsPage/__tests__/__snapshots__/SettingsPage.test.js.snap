// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPage should render 1`] = `
<PageLayout
  beforeToolbarComponent={null}
  breadcrumbOptions={null}
  customBreadcrumbs={null}
  header="Settings"
  isLoading={false}
  onBookmarkClick={[Function]}
  onSearch={[Function]}
  searchProps={
    Object {
      "autocomplete": Object {
        "id": "searchBar",
        "searchQuery": "",
        "url": "settings/auto_complete_search",
        "useKeyShortcuts": true,
      },
      "bookmarks": Object {
        "canCreate": true,
        "documentationUrl": "4.1.5Searching",
        "url": "/api/bookmarks",
      },
      "controller": "settings",
    }
  }
  searchQuery=""
  searchable={true}
  toastNotifications={null}
  toolbarButtons={null}
>
  <Component
    errorMsg={Object {}}
    groupedSettings={Object {}}
    hasData={false}
    hasError={false}
    isLoading={true}
  />
</PageLayout>
`;
